File: spu/state_tracker/state_current.py
import sys;

sys.path.append( '../../packer' )
from pack_currenttypes import *;

print '''
#include "state/cr_currentpointers.h"
#include "cr_glstate.h"

#include <stdio.h>

static double __read_double( void *src )
{
    unsigned int *ui = (unsigned int *) src;
    double d;
    ((unsigned int *) &d)[0] = ui[0];
    ((unsigned int *) &d)[1] = ui[1];
    return d;
}
'''

for k in gltypes.keys():
	for i in range(1,5):
		print 'void __convert_%s%d (GLfloat *dst, %s *src) {' % (k,i,gltypes[k]['type'])
		if k == 'd':
		  for j in range(1,i-1):
		    print '\t*dst++ = (GLfloat) __read_double(src++);'
		  print '\t*dst = (GLfloat) __read_double(src);'
		else:
		  for j in range(1,i-1):
		    print '\t*dst++ = (GLfloat) *src++;';
		  print '\t*dst = (GLfloat) *src;';
		print '}\n';

scale = {
	'ub' : 'GL_MAXUBYTE',
	'b'  : 'GL_MAXBYTE',
	'us' : 'GL_MAXUSHORT',
	's'  : 'GL_MAXSHORT',
	'ui' : 'GL_MAXUINT',
	'i'  : 'GL_MAXINT',
	'f'  : '',
	'd'  : ''
}

for k in gltypes.keys():
	if k != 'f' and k != 'd' and k != 'l':
		for i in range(1,5):
			print 'void __convert_rescale_%s%d (GLfloat *dst, %s *src) {' % (k,i,gltypes[k]['type'])
			for j in range(1,i-1):
				print '\t*dst++ = ((GLfloat) *src++) / %s;' % scale[k]
			print '\t*dst = ((GLfloat) *src) / %s;' % scale[k]
			print '}\n'

print '''

void __convert_boolean (GLboolean *dst, GLboolean *src) {
	*dst = *src;
}

typedef void (*convert_func) (GLfloat *, unsigned char *);
'''

for k in current_fns.keys():
	name = k
	name = '%s%s' % (k[:1].lower(),k[1:])
	ucname = k.upper()
	num_members = len(current_fns[k]['default']) + 1

	print '#define VPINCH_CONVERT_%s(op,data,dst) \\' % ucname
	print '{\\'
	print '\tGLfloat vdata[%d] = {' % num_members,

## Copy dst data into vdata
	i = 0;
	for defaultvar in current_fns[k]['default']:
		print '%d' % defaultvar,
		if i != num_members:
			print ',',
		i += 1
	print '};\\'

	print '\tswitch (op) { \\'
	for type in current_fns[k]['types']:
		for size in current_fns[k]['sizes']:
			uctype = type.upper()
			if ucname == 'EDGEFLAG':
				print '\tcase CR_%s_OPCODE: \\' % ucname
			else:
				print '\tcase CR_%s%d%s_OPCODE: \\' % (ucname,size,uctype)
			
			if (ucname == 'COLOR' or ucname == 'NORMAL') and type != 'f' and type != 'd':
				print '\t\t__convert_rescale_%s%d (vdata, (%s *) (data)); \\' % (type,size,gltypes[type]['type'])
			else:
				print '\t\t__convert_%s%d (vdata, (%s *) (data)); \\' % (type,size,gltypes[type]['type'])
			print '\t\tbreak; \\'

	print '\tdefault: \\'
	print '\t\tcrSimpleError ( "Unknown opcode in VPINCH_CONVERT_%s" ); \\' % ucname
	print '\t}\\'

	i = 0
	for member in current_fns[k]['members']:
		print '\t(dst).%s = vdata[%d];\\' % (member,i)
		i += 1

	print '}\n'

print '''

void crStateCurrentRecover( void )
{
	unsigned char *v;
	convert_func convert=NULL;
	CRContext *g = GetCurrentContext();
	CRCurrentState *c = &(g->current);
	CRStateBits *b = GetCurrentBits();
	CRCurrentBits *cb = &(b->current);
	GLbitvalue nbitID = g->neg_bitid;
	static const GLcolorf color_default			= {0.0f, 0.0f, 0.0f, 1.0f};
	static const GLtexcoordf texCoord_default	= {0.0f, 0.0f, 0.0f, 1.0f};
	static const GLvectorf normal_default		= {0.0f, 0.0f, 0.0f, 1.0f};
	static const GLfloat index_default			= 0.0f;
	static const GLboolean edgeFlag_default		= GL_TRUE;
	GLnormal_p		*normal		= &(c->current->normal);
	GLcolor_p		*color		= &(c->current->color);
	GLtexcoord_p	*texCoord	= &(c->current->texCoord);
	GLindex_p		*index		= &(c->current->index);
	GLedgeflag_p	*edgeFlag	= &(c->current->edgeFlag);

	/* Save pre state */
	c->normalPre = c->normal;
	c->colorPre = c->color;
	c->texCoordPre = c->texCoord;
	c->indexPre = c->index;
	c->edgeFlagPre = c->edgeFlag;

'''

for k in current_fns.keys():
	print '\t/* %s */' % k
	print '\tv=NULL;'
	name = '%s%s' % (k[:1].lower(),k[1:])

	for type in current_fns[k]['types']:
		for size in current_fns[k]['sizes']:
			ptr = '%s->%s%d' % (name, type, size )
			print '\tif (v < %s)' % ptr
			print '\t{'
			print '\t\tv = %s;' % ptr
			if (k == 'Color' or k == 'Normal') and type != 'f' and type != 'd' and type != 'l':
				print '\t\tconvert = (convert_func) __convert_rescale_%s%d;' % (type,size)
			else:
				print '\t\tconvert = (convert_func) __convert_%s%d;' % (type,size)
			print '\t}'
	print ''
	print '\tif (v != NULL) {'
	print '\t\tc->%s = %s_default;' % (name,name)
	if k == 'EdgeFlag':
		print '\t\t__convert_boolean (&(c->%s), v);' % name
	elif k == 'Index':
		print '\t\tconvert(&(c->%s), v);' % name
	elif k == 'Normal':
		print '\t\tconvert(&(c->%s.x), v);' % name
	elif k == 'TexCoord':
		print '\t\tconvert(&(c->%s.s), v);' % name
	elif k == 'Color':
		print '\t\tconvert(&(c->%s.r), v);' % name
	print '\t\tcb->%s = nbitID;' % name
	print '\t\tcb->dirty = nbitID;'
	print '\t}'
	print '\t%s->ptr = v;' % name
print '}'

File: spu/state_tracker/state_current.c
/* Copyright (c) 2001, Stanford University
 * All rights reserved
 *
 * See the file LICENSE.txt for information on redistributing this software.
 */

#include <stdio.h>
#include <stdlib.h>
#include "cr_glstate.h"
#include "cr_glwrapper.h"
#include "state_internals.h"

void crStateCurrentInit( CRCurrentState *c )
{
	GLvectorf	default_normal     = {0.0f, 0.0f, 1.0f, 1.0f};
	GLcolorf	default_color	     = {1.0f, 1.0f, 1.0f, 1.0f};
	GLcolorf	default_secondaryColor = {0.0f, 0.0f, 0.0f, 0.0f};
	GLtexcoordf default_texcoord = {0.0f, 0.0f, 0.0f, 1.0f};
	GLvectorf default_rasterpos  = {0.0f, 0.0f, 0.0f, 1.0f};
	int i;

	c->color	= default_color;
	c->secondaryColor = default_secondaryColor;
	c->index	= 1.0f;
	for (i = 0 ; i < CR_MAX_TEXTURE_UNITS ; i++)
	{
		c->texCoord[i] = default_texcoord;
	}
	c->normal	= default_normal;

	c->rasterPos = default_rasterpos;
	c->rasterPosPre = c->rasterPos;

	c->rasterDistance = 0.0f;
	c->rasterColor = default_color;
	c->rasterSecondaryColor = default_secondaryColor;
	c->rasterTexture = default_texcoord;
	c->rasterValid = GL_TRUE;
	c->rasterIndex = 1.0f;

	c->edgeFlag = GL_TRUE;
	c->normalize = GL_FALSE;

	c->inBeginEnd = GL_FALSE;
	c->beginEndNum = 0;
#if 0
	c->beginEndMax = cfg->beginend_max;
#endif
	c->mode = 0x10; /* Undefined Mode */
	c->flushOnEnd = 0;

#if 0
	c->current = cfg->current;
#endif
}

void STATE_APIENTRY crStateColor3f( GLfloat r, GLfloat g, GLfloat b )
{
	crStateColor4f(r, g, b, 1.0F);
}

void STATE_APIENTRY crStateColor3fv( const GLfloat *color )
{
	crStateColor4f( color[0], color[1], color[2], 1.0F );
}

void STATE_APIENTRY crStateColor4f( GLfloat red, GLfloat green, GLfloat blue, GLfloat alpha )
{
	CRContext *g = GetCurrentContext();
	CRCurrentState *c = &(g->current);
	CRStateBits *sb = GetCurrentBits();
	CRCurrentBits *cb = &(sb->current);

	FLUSH();

        c->color.r = red;
        c->color.g = green;
        c->color.b = blue;
        c->color.a = alpha;

        DIRTY(cb->dirty, g->neg_bitid);
        DIRTY(cb->color, g->neg_bitid);
}

void STATE_APIENTRY crStateColor4fv( const GLfloat *color )
{
	crStateColor4f( color[0], color[1], color[2], color[3] );
}

void crStateSetCurrentPointers( CRContext *ctx, CRCurrentStatePointers *current )
{
	CRCurrentState *c = &(ctx->current);
	c->current = current;
}

void STATE_APIENTRY crStateBegin( GLenum mode )
{
	CRContext *g = GetCurrentContext();
	CRCurrentState *c = &(g->current);

	if (mode > GL_POLYGON)
	{
		crStateError(__LINE__, __FILE__, GL_INVALID_ENUM, "Begin called with invalid mode: %d", mode);
		return;
	}

	if (c->inBeginEnd)
	{
		crStateError( __LINE__, __FILE__, GL_INVALID_OPERATION, "glBegin called inside Begin/End");
		return;
	}

	c->inBeginEnd = GL_TRUE;
	c->mode = mode;
	c->beginEndNum++;
}

void STATE_APIENTRY crStateEnd( void )
{
	CRContext *g = GetCurrentContext();
	CRCurrentState *c = &(g->current);

	if (!c->inBeginEnd)
	{
		crStateError( __LINE__, __FILE__, GL_INVALID_OPERATION, "glEnd called outside Begin/End" );
		return;
	}


	c->inBeginEnd = GL_FALSE;
}

void crStateCurrentSwitch( GLuint maxTextureUnits,
						   CRCurrentBits *c, GLbitvalue *bitID,
						   CRCurrentState *from, CRCurrentState *to )
{
	unsigned int i,j;
	GLbitvalue nbitID[CR_MAX_BITARRAY];

	for (j=0;j<CR_MAX_BITARRAY;j++)
		nbitID[j] = ~bitID[j];

	if (CHECKDIRTY(c->enable, bitID)) {
		if (from->normalize != to->normalize) {
			if (to->normalize == GL_TRUE)
				diff_api.Enable(GL_NORMALIZE);
			else
				diff_api.Disable(GL_NORMALIZE);
			FILLDIRTY(c->enable);
			FILLDIRTY(c->dirty);
		}
		INVERTDIRTY(c->enable, nbitID);
	}

	if (CHECKDIRTY(c->raster, bitID)) {
		if (to->rasterValid) {
			if (to->rasterPosPre.x != from->rasterPos.x ||
				to->rasterPosPre.y != from->rasterPos.y) {
					GLvectorf p;
					p.x = to->rasterPosPre.x - from->rasterPos.x;
					p.y = to->rasterPosPre.y - from->rasterPos.y;
					diff_api.Bitmap(0, 0, 0.0f, 0.0f, p.x, p.y, 0);
					FILLDIRTY(c->raster);
					FILLDIRTY(c->dirty);
			}
		}
		INVERTDIRTY(c->raster, nbitID);
	}

	/* Vertex Current State Switch Code */

	/* Its important that we don't do a value check here because
	** current may not actaully have the correct values, I think...
	** We also need to restore the current state tracking pointer
	** since the packing functions will set it.
	*/

	/* NEED TO FIX THIS!!!!!! */
	if (CHECKDIRTY(c->color, bitID)) {
           if (COMPARE_COLOR(from->color,to->color)) {
			diff_api.Color4fv ((GLfloat *) &(to->color));
			FILLDIRTY(c->color);
			FILLDIRTY(c->dirty);
		}
		INVERTDIRTY(c->color, nbitID);
	}

	/* NEED TO FIX THIS, ALSO?!!!!! */
#ifdef CR_EXT_secondary_color
	if (CHECKDIRTY(c->secondaryColor, bitID)) {
		if (COMPARE_COLOR(from->secondaryColor,to->secondaryColor)) {
			diff_api.SecondaryColor3fvEXT ((GLfloat *) &(to->secondaryColor));
			FILLDIRTY(c->secondaryColor);
			FILLDIRTY(c->dirty);
		}
		INVERTDIRTY(c->secondaryColor, nbitID);
	}
#endif

	if (CHECKDIRTY(c->index, bitID)) {
		if (to->index != from->index) {
			diff_api.Indexf (to->index);
			FILLDIRTY(c->index);
			FILLDIRTY(c->dirty);
		}
		INVERTDIRTY(c->index, nbitID);
	}

	if (CHECKDIRTY(c->normal, bitID)) {
		if (COMPARE_VECTOR (from->normal, to->normal)) {
			diff_api.Normal3fv ((GLfloat *) &(to->normal.x));
			FILLDIRTY(c->normal);
			FILLDIRTY(c->dirty);
		}
		INVERTDIRTY(c->normal, nbitID);
	}

	for (i = 0; i < maxTextureUnits; i++)
	{
		if (CHECKDIRTY(c->texCoord[i], bitID)) {
			if (COMPARE_TEXCOORD (from->texCoord[i], to->texCoordPre[i])) {
				diff_api.MultiTexCoord4fvARB (i+GL_TEXTURE0_ARB, (GLfloat *) &(to->texCoord[i].s));
				FILLDIRTY(c->normal);
				FILLDIRTY(c->dirty);
			}
			INVERTDIRTY(c->texCoord[i], nbitID);
		}
	}

	INVERTDIRTY(c->dirty, nbitID);
}

void crStateCurrentDiff (CRCurrentBits *c, GLbitvalue *bitID,
					 CRCurrentState *from, CRCurrentState *to)
{
	int i,j;
	GLbitvalue nbitID[CR_MAX_BITARRAY];

	for (j=0;j<CR_MAX_BITARRAY;j++)
		nbitID[j] = ~bitID[j];

	if (CHECKDIRTY(c->enable, bitID)) {
		if (from->normalize != to->normalize) {
			if (to->normalize == GL_TRUE)
				diff_api.Enable(GL_NORMALIZE);
			else
				diff_api.Disable(GL_NORMALIZE);
			from->normalize = to->normalize;
		}
		INVERTDIRTY(c->enable, nbitID);
	}

	if (CHECKDIRTY(c->raster, bitID)) {
		from->rasterValid = to->rasterValid;
		if (to->rasterValid) {
			if (to->rasterPosPre.x != from->rasterPos.x ||
				to->rasterPosPre.y != from->rasterPos.y) {
					GLvectorf p;
					p.x = to->rasterPosPre.x - from->rasterPos.x;
					p.y = to->rasterPosPre.y - from->rasterPos.y;
					diff_api.Bitmap(0, 0, 0.0f, 0.0f, p.x, p.y, 0);
			}
			from->rasterPos = to->rasterPos;
		}
		INVERTDIRTY(c->raster, nbitID);
	}

	/* Vertex Current State Sync Code */
	/* Some things to note here:
	** 1) Compare is done against the pre value since the
	**    current value includes the geometry info.
	** 2) Update is done with the current value since
	**    the server will be getting the geometry block
	** 3) Copy is done outside of the compare to ensure
	**    that it happens.
	*/
	if (CHECKDIRTY(c->color, bitID)) {
		if (COMPARE_COLOR(from->color,to->colorPre)) {
			diff_api.Color4fv ((GLfloat *) &(to->colorPre.r));
		}
		from->color = to->color;
		INVERTDIRTY(c->color, nbitID);
	}

#ifdef CR_EXT_secondary_color
	if (CHECKDIRTY(c->secondaryColor, bitID)) {
		if (COMPARE_COLOR(from->secondaryColor,to->secondaryColorPre)) {
			diff_api.SecondaryColor3fvEXT ((GLfloat *) &(to->secondaryColorPre.r));
		}
		from->secondaryColor = to->secondaryColor;
		INVERTDIRTY(c->secondaryColor, nbitID);
	}
#endif

	if (CHECKDIRTY(c->index, bitID)) {
		if (from->index != to->indexPre) {
			diff_api.Indexf (to->index);
		}
		from->index = to->index;
		INVERTDIRTY(c->index, nbitID);
	}

	if (CHECKDIRTY(c->normal, bitID)) {
		if (COMPARE_VECTOR (from->normal, to->normalPre)) {
			diff_api.Normal3fv ((GLfloat *) &(to->normalPre.x));
		}
		from->normal = to->normal;
		INVERTDIRTY(c->normal, nbitID);
	}

	for ( i = 0 ; i < CR_MAX_TEXTURE_UNITS ; i++)
	{
		if (CHECKDIRTY(c->texCoord[i], bitID)) {
			if (COMPARE_TEXCOORD (from->texCoord[i], to->texCoordPre[i])) {
				diff_api.MultiTexCoord4fvARB (GL_TEXTURE0_ARB + i, (GLfloat *) &(to->texCoordPre[i].s));
			}
			from->texCoord[i] = to->texCoord[i];
			INVERTDIRTY(c->texCoord[i], nbitID);
		}
	}

	if (CHECKDIRTY(c->edgeFlag, bitID)) {
		if (from->edgeFlag != to->edgeFlagPre) {
			diff_api.EdgeFlag (to->edgeFlagPre);
		}
		from->edgeFlag = to->edgeFlag;
		INVERTDIRTY(c->edgeFlag, nbitID);
	}

	INVERTDIRTY(c->dirty, nbitID);
}
